storage/innobase/lock/lock0lock.cc

```diff
@@ -4680,73 +4800,119 @@ and release possible other transactions waiting because of these locks. */
 void lock_release_on_prepare(trx_t *trx)
 {
   trx->set_skip_lock_inheritance();
+  bool unlock_unmodified=
+#ifdef HAVE_REPLICATION
+      thd_is_slave(trx->mysql_thd) ||
+#endif
+      innodb_enable_xap_unlock_unmodified_for_primary_debug;
 
+#ifdef ENABLED_DEBUG_SYNC
+  DBUG_EXECUTE_IF("skip_lock_release_on_prepare_try", goto skip_try;);
+#endif
   for (ulint count= 5; count--; )
-    if (lock_release_on_prepare_try(trx))
+    if (lock_release_on_prepare_try(trx, unlock_unmodified))
       return;
+#ifdef ENABLED_DEBUG_SYNC
+skip_try:
+#endif
 
   mtr_t mtr;
-  rec_offs offsets_[REC_OFFS_NORMAL_SIZE];
+  /* Reserve enough offsets for the key and PRIMARY KEY. */
+  rec_offs offsets_[REC_OFFS_HEADER_SIZE + 2 * MAX_REF_PARTS + 1];
   rec_offs *offsets= offsets_;
   mem_heap_t *heap= nullptr;
 
   rec_offs_init(offsets_);
```

Comment:
Both `offsets` and `heap` should be removed. A local variable `heap` should be declared in `lock_rec_unlock_unmodified()`, which this function would invoke on `offsets_`.